File: FE/lib/services/order_service.dart
import 'package:soft/services/api_service.dart';

class OrderService {
  final ApiService _apiService =
      ApiService(); // sử dụng chung ApiService

  Future<Map<String, dynamic>> createOrder(
    Map<String, dynamic> data,
  ) async {
    return await _apiService.post('/order/create', data);
  }

  Future<Map<String, dynamic>> getAllOrder() async {
    return await _apiService.get('/order/getAll');
  }

  Future<Map<String, dynamic>> delete(String id) async {
    return await _apiService.delete('/order/delete/$id');
  }

  Future<Map<String, dynamic>> update(
    String id,
    Map<String, dynamic> data,
  ) async {
    return await _apiService.put('/order/update/$id', data);
  }

  Future<Map<String, dynamic>> getByUser() async {
    return await _apiService.get('/order/getByUser');
  }

  Future<Map<String, dynamic>> getbyId(String id) async {
    return await _apiService.get('/order/getById/$id');
  }

  Future<Map<String, dynamic>> scanWork(
    Map<String, dynamic> data,
  ) async {
    return await _apiService.post('/order/scanWork', data);
  }
}

File: WebApp/Frontend/src/services/orderService.ts
import { useQuery } from "@tanstack/react-query";
import api from "../config/api.config";
import { Device, Location, Material, Order } from "../types";
import { Dayjs } from "dayjs";

const OrderService = {
    getAll: async (params?: Record<string, any>): Promise<any> => {
        const res = await api.get('/orders', { params });
        return res.data
    },
    getAllDispatcher: async (params?: Record<string, any>): Promise<any[]> => {
        const res = await api.get('/orders', { params });
        return res.data.data
    },
    getByUser: async (params?: Record<string, any>): Promise<any> => {
        const res = await api.get('/orders/user', { params });
        return res.data
    },
    create: async (data: Partial<Order>): Promise<any> => {
        const res = await api.post('/orders', data);
        return res.data
    },
    update: async (data: Partial<Order>): Promise<any> => {
        const res = await api.put(`/orders/${data?._id}`, data);
        return res.data
    },
    delete: async (ids: string[]): Promise<any> => {
        const res = await api.delete(`/orders`, { data: { ids } });
        return res.data.message
    },
    importFile: async (
        formData: FormData,
        onProgress?: (percent: number) => void
    ) => {
        const res = await api.post("/orders/importFile", formData, {
            headers: { "Content-Type": "multipart/form-data" },
            onUploadProgress: (e) => {
                if (!onProgress) return;
                const total = e.total ?? 1;
                const percent = Math.round((e.loaded * 100) / total);
                onProgress(percent);
            },
        });
        return res.data.message
    },
    exportFile: async (
        selectedOrders: any[]
    ) => {
        const res = await api.post('/exports/order/bulk', { ids: selectedOrders.map(o => o._id) }, {
            responseType: 'blob',
        });
        const blob = new Blob([res.data], {
            type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
        });

        const url = window.URL.createObjectURL(blob);
        const link = document.createElement('a');
        link.href = url;
        link.setAttribute('download', `*.xlsx`);

        document.body.appendChild(link);
        link.click();
        link.parentNode?.removeChild(link);
        window.URL.revokeObjectURL(url);
    },
    exportFileList: async (
        selectedOrders: any[],
        isSelectedAll: boolean,
        status: string
    ) => {
        const res = await api.post('/orders/exportFile/bulk', {
            ids: selectedOrders.map(o => o._id),
            isSelectedAll,
            status
        }, {
            responseType: 'blob',
        });
        const blob = new Blob([res.data], {
            type: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet',
        });

        const url = window.URL.createObjectURL(blob);
        const link = document.createElement('a');
        link.href = url;
        link.setAttribute('download', `*.xlsx`);

        document.body.appendChild(link);
        link.click();
        link.parentNode?.removeChild(link);
        window.URL.revokeObjectURL(url);
    }
};

export default OrderService;